### ISSUE TITLE
Tiny things for collapsed toots

### ISSUE THREAD

**Issue by marrus-sh** [glitch-soc/mastodon#22]
I don't wanna make a bunch of issues and milestones lol but these are things that would be nice to have with collapsed toots:

- [x] Basic implementation added in 93c52301ade71fb685d937b90451974d03d710af with styling improvements in 4bd7482a7a010f5f8bdcd53869f871bf93ed1fc0, 40ecbfd4a93d88d1483269755692df3b2d93a52b, and 2513d92c54aaf30da3b658acdca8a70fb0a85a28
- [x] Prevent reblogs from breaking the website via collapsed toots
    - Implemented in 67adbcc60cee7915f8e59256a0eaf6d42a03a559
- [x] Update height of toot on collapsing
    - Implemented in 4fe45dda9a046fc18e20c3d3568348b6b5aefa07
    - Fixed with respect to reblogs through 36805a39db2bd1545d58a6731f988995dbf98c11
- [x] Show media as a filtered background on collapsed toots
    - For toots with multiple media attachments, the first would be used
    - For gifv and video, the thumbnail would be used
    - Implemented in efacfec3ed37c200bf5bad05b59e04a9f97d0d08
    - [x] This should be A Setting
    - Setting implemented in 595c6de32c052f2a5e21307a0e43a6bd5d4b5c88
- [x] Include a media icon next to the collapse button
    - Implemented in 4bc237fcfe2838edb69baec87bf95ece4fc26933 and recolored in ddc6b8591230f0d29b1bc6d881c62a7524a16ab0
    - [x] (Optional) Also include a number specifying the number of media attachments
        - No number is provided, but photosets are given a different icon in 4bc237fcfe2838edb69baec87bf95ece4fc26933
- [x] ~Turn off the `[Show Less/More]` button on collapsed toots~ Make expanding content warning also expand the toot
    - Implemented in d6e3918d924436529f9bce9bfe54eb0b6619c637
    - Reverted in 6a50e730896f599f16961c118126b805e26e2f53, which is a better solution
- [x] Click anywhere to expand a toot
    - Implemented for status text in e60f27d6490b21a75cf764a22a011c63ce8532a7 and a520b118e4653aa35a74eba6e1662dfc98fcd1cc with a slight appearance fix in aeff898137b0629dc71cd6f537f3a060e15a8e40
    - Implemented for links in 5df7bc3a8bee95a751e7d55b2513653a2e1e9a4e
    - Implemented for header text/avatars in 59503a88ae96f554eab3c33f5760e98808ee135a
- [x] Make notification toots collapsed-by-default
    - Implemented in c806fef865a9364bef59a6d90e99166bb9e9546c
    - [x] It would be nice if this was A Setting
    - Setting implemented in 595c6de32c052f2a5e21307a0e43a6bd5d4b5c88
- [x] Toots don't seem to auto-collapse in Notifications, Conversations, etc (!?)
    - Hopefully fixed by c806fef865a9364bef59a6d90e99166bb9e9546c
- [x] Setting to auto-collapse toots with replies or toots with media or all toots
    - Implemented as an App Setting in 595c6de32c052f2a5e21307a0e43a6bd5d4b5c88
- [x] Don't include media when calculating whether or not to collapse a toot, so that artists don't get all their toots auto-collapsed
    - Fixed as a part of ec2daae71ca03b414fa41a07e13e9653b0dff0a4
- [x] Improve styling of collapsed toots so that links/headers don't look clickable
    - Done for headers in 99889ea57d04646d593dc5e9d0b1c8fe97856874, but still needs to be done for links
    - Done for links in bba75c15f1f7a6878db730fdb3930d6b7a823da7
- [x] Change the appearance of the "boosted" wrapper to work better with collapsed toots
    - Boosted wrapper placed inside of statuses with bba75c15f1f7a6878db730fdb3930d6b7a823da7
- [x] Improve the appearance of collapsed notifications with backgrounds
    - Fixed via better filter in bba75c15f1f7a6878db730fdb3930d6b7a823da7
- [x] Revise the background filter on collapsed toots to more closely reflect the filters used elsewhere on the site; for example, on static user pages
    - Done in bba75c15f1f7a6878db730fdb3930d6b7a823da7
- [x] Shift+click to collapse a toot
    - Implemented in bba75c15f1f7a6878db730fdb3930d6b7a823da7
- 🙅 Rename "uncollapse" to "expand" lol
    - Probably a __WONTFIX__ since Expand/Unexpand is CW terminology
- 🙅 Include a conversation icon next to the collapse button, maybe with the number of toots in the conversation, for replies/threads
    - I don't know if we can give the actual length of the conversation (in a twitter-esque “X replies” manner
    - But we *can* say how many statuses with that `conversation_id` we've encountered
    - Probably a __WONTFIX__ in favor of actual collapsed conversations
- 🙅 Button to collapse-all or uncollapse-all toots in a timeline
    - May depend on #27
    - Can be made its own issue probably

**Comment by marrus-sh** [comment]
Please ignore this fake code I'm just writing it down so I don't forget. Here's how you implement the collapse-all and uncollapse-all button internally; hopefully also the backbone for implementing other kinds of automated collapsing.
```js
// whatever thingy has the button
// state.collapse_all is normally `undefined`
handleCollapseAll : () => { // onClick for collapse all button
  this.setState({ collapse_all: true });
}
handleUncollapseAll : () => { // onClick for expand all button
  this.setState({ collapse_all: false });
}
componentDidUpdate : () => { // Resets `collapse_all` after every update
  if (this.state.collapse_all != null) this.setState({ collapse_all: undefined });
}
```
```js
// status.js
componentWillReceiveProps: (nextProps) => { // the `collapse` prop will be `true` to collapse, `false` to expand, `undefined` to do nothing
  if (nextProps.collapse != null) this.setState({ isCollapsed: !!nextProps.collapse });
}
```
